system/jlib/jstring.cpp

```diff
@@ -939,25 +939,73 @@ StringBuffer &replaceString(StringBuffer & result, size_t lenSource, const char
 {
     if (lenSource)
     {
-        size_t left = lenSource;
-        while (left >= lenOldStr)
+        if (lenOldStr)
         {
-            if (memcmp(source, oldStr, lenOldStr)==0)
+            // Scan for match on first character
+            size_t offset = 0;
+            size_t left = lenSource;
+            while (left >= lenOldStr)
             {
-                result.append(lenNewStr, newStr);
-                source += lenOldStr;
-                left -= lenOldStr;
+                if (source[offset] != oldStr[0])
+                {
+                    offset++;
+                    left--;
+                }
+                else
+                    break;
             }
-            else
+
+            // Reserve space when the new string is longer than the old string
+            unsigned steps = lenSource - lenOldStr - offset + 1;
+            unsigned maxResultLength = lenNewStr > lenOldStr ? lenSource + steps * (lenNewStr - lenOldStr) : lenSource;
+            result.ensureCapacity(maxResultLength);
+
+            // Add any characters before the match
+            result.append(offset, source);
+
+            size_t unmatchedChars = 0;
+            while (left >= lenOldStr)
             {
-                result.append(*source);
-                source++;
-                left--;
+                if (memcmp(source + offset, oldStr, lenOldStr)==0)
+                {
+                    if (unmatchedChars)
+                    {
+                        // Copy mismatched characters in blocks
+                        result.append(unmatchedChars, source + offset - unmatchedChars);
+                        unmatchedChars = 0;
+                    }
+                    result.append(lenNewStr, newStr);
+                    offset += lenOldStr;
+                    left -= lenOldStr;
+                }
+                else
+                {
+                    unmatchedChars++;
+                    offset++;
+                    left--;
+
+                    // Skip ahead to the next possible match
+                    while (left >= lenOldStr)
```

Comment:
This is essentially repeating the code at the head of the function.  Better to restructure so that the code only appears once.  E.g.

```
            size_t offset = 0;
            size_t lastCopied = 0;
            size_t maxOffset = lenSource-lenOldStr;
            char firstChar = oldStr[0];
            result.ensureCapacity(lenSource);  // see comment above.
            while (offset <= maxOffset)
            {
                if (source[offset] == firstChar && (memcmp(source+offset, oldStr, lenOldStr) == 0)
                {
                        if (offset != lastCopied)
                            result.append(offset-lastCopied, source + offset);
                        result.append(lenNewStr, newStr);
                        lastCopied = offset + lenOldStr;
                        offset += lenOldStr;
                     }
                 }
                 else
                     offset += 1;
            }
            result.append(lenOldStr - lastCopied, oldStr + lastCopied);
      }
      else....
```
NOTE: This code has not been tested, it was typed in here.

My aim is to make the code as simple (and as clear) as possible, because that is likely to produce the most efficient code.  For example when a character does not match this is only updating a single variable, rather than 3 variables.  It has avoided duplicating the code for scanning.

A further improvement for code that is really time critical is to give the compiler hints about how likely something is.  We have macros likely() and unlikely() which pass through information to the compiler.  So even better:

```
                if (unlikely(source[offset] == firstChar) && 
                    unlikely(memcmp(source+offset, oldStr, lenOldStr) == 0)
```
I am fairly sure this is the default bahavior of the compiler in this case, but it is better to be explicit.

Reply:
Restructured code and added unlikely macros.